internal/jimm/model.go

```diff
@@ -1347,3 +1351,67 @@ func (j *JIMM) ChangeModelCredential(ctx context.Context, user *openfga.User, mo
 
 	return nil
 }
+
+// ListModels list the models that the user has access to. It intentionally excludes the
+// controller model as this call is used within the context of login and register commands.
+func (j *JIMM) ListModels(ctx context.Context, user *openfga.User) ([]base.UserModel, error) {
+	const op = errors.Op("jimm.ListModels")
+	zapctx.Info(ctx, string(op))
+
+	// Get models uuids user has access to
+	uuids, err := user.ListModels(ctx, ofganames.ReaderRelation)
+	if err != nil {
+		return nil, errors.E(op, err, "failed to list models")
+	}
+
+	// Get the models themselves
+	models, err := j.DB().GetModelsByUUID(ctx, uuids)
+	if err != nil {
+		return nil, errors.E(op, err, "failed to get models by uuid")
+	}
+
+	// Find the controllers these models reside on and remove duplicates
+	var controllers []dbmodel.Controller
+	seen := make(map[uint]bool)
+	for _, model := range models {
+		if seen[model.ControllerID] {
+			continue
+		}
+		seen[model.ControllerID] = true
+		controllers = append(controllers, model.Controller)
+	}
+
+	// Call controllers for their models. We always call as admin, and we're
+	// filtering ourselves. We do this rather than send the user to be 100%
+	// certain that the models do belong to user according to OpenFGA. We could
+	// in theory rely on Juju correctly returning the models (by owner), but this
+	// is more reliable.
+	var userModels []base.UserModel
+	var mutex sync.Mutex
+	err = j.forEachController(ctx, controllers, func(_ *dbmodel.Controller, api API) error {
+		ums, err := api.ListModels(ctx)
+		if err != nil {
+			return err
+		}
+		mutex.Lock()
+		defer mutex.Unlock()
+
+		// Filter the models returned according to the uuids
+		// returned from OpenFGA for read access.
+		//
+		// NOTE: We skip controller models as ListModels is used for login and register.
+		// The models returned are stored locally and used for reference. In the case of JIMM,
+		// we do not want to show the controller models.
+		for _, m := range ums {
+			if slices.Contains(uuids, m.UUID) {
+				userModels = append(userModels, m)
```

Comment:
i dont think you can return it right away from the controller, but rather override the `OwnerTag` from our db

Reply:
I'm not sure what you mean here?

Reply:
here we return an `OwnerTag`. In your implementation we return the `OwnerTag` coming from the controllers, in most of the other model's method we override the `OwnerTag` with the one in our database.
Are they always the same? https://github.com/canonical/jimm/blob/19a40d7c399638e56dc33bc5a876b0aae7a56375/internal/dbmodel/model.go#L152

Reply:
another example:
https://github.com/canonical/jimm/blob/19a40d7c399638e56dc33bc5a876b0aae7a56375/internal/dbmodel/model.go#L129

Reply:
This is models the user has *access* to, not that they necessarily *own* the model?

Reply:
As disussed is OK, resolving...